File: hoover.ts
/* Copyright (c) 2021-2023 Richard Rodger, MIT License */

/*
support regexp
indent removal
more rule cases
*/

// TODO: line continuation ("\" at end) should be a feature of standard JSONIC strings

import {
  Jsonic,
  Rule,
  RuleSpec,
  Plugin,
  Config,
  Options,
  Lex,
  Point,
  Context,
  MakeLexMatcher,
  makeStringMatcher,
  makePoint,
  Token,
  AltAction,
} from '@jsonic/jsonic-next'

type HooverOptions = {
  block: {
    [name: string]: {
      start?: {
        fixed?: string | string[]
        consume?: null | boolean | string[] // explicit false to turn off
      }
      end?: {
        fixed?: string | string[]
        consume?: null | boolean | string[] // explicit false to turn off
      }
      escapeChar?: string
      escape?: {
        [char: string]: string
      }
      allowUnknownEscape: boolean
      preserveEscapeChar: boolean
      trim: boolean
    }
  }
  lex?: {
    order?: number
  }
  action?: AltAction
}

const Hoover: Plugin = (jsonic: Jsonic, options: HooverOptions) => {
  const { entries } = jsonic.util

  let blocks = entries(options.block).map((entry: any[]) => ({
    allowUnknownEscape: true,
    preserveEscapeChar: false,
    token: '#HV',
    ...entry[1],
    name: entry[0],
  }))

  let tokenMap: any = {}

  for (let block of blocks) {
    // Create a hoover token
    block.TOKEN = jsonic.token(block.token)

    if (!tokenMap[block.token]) {
      jsonic.rule('val', (rs) => {
        rs.open({
          s: [block.TOKEN],
          a: options.action,
        })
      })
    }

    tokenMap[block.token] = block.TOKEN
  }

  let makeHooverMatcher: MakeLexMatcher = (cfg: Config, _opts: Options) => {
    return function hooverMatcher(lex: Lex) {
      for (let block of blocks) {
        // TODO: Point.clone ?
        const hvpnt = makePoint(lex.pnt.len, lex.pnt.sI, lex.pnt.rI, lex.pnt.cI)

        let startResult = matchStart(lex, hvpnt, block)

        if (startResult.match) {
          let result = parseToEnd(lex, hvpnt, block, cfg)

          if (result.done) {
            let tkn = lex.token(
              block.TOKEN,
              result.val,
              lex.src.substring(lex.pnt.sI, hvpnt.sI),
              hvpnt,
            )
            tkn.use = { block: block.name }

            lex.pnt.sI = hvpnt.sI
            lex.pnt.rI = hvpnt.rI
            lex.pnt.cI = hvpnt.cI

            return tkn
          } else {
            return result.bad || lex.bad('invalid_text', lex.pnt.sI, hvpnt.sI)
          }
        }
      }

      return undefined
    }
  }

  jsonic.options({
    lex: {
      match: {
        hoover: { order: options.lex?.order, make: makeHooverMatcher },
      },
    },
  })
}

function matchStart(
  lex: Lex,
  hvpnt: Point,
  block: any,
): {
  match: boolean
  start?: string
} {
  let src = lex.src

  let sI = hvpnt.sI // Current point in src
  let rI = hvpnt.rI // Current row
  let cI = hvpnt.cI // Current column

  let start = block.start || {}
  let rulespec = start.rule || {}
  let matchRule: null | boolean = null

  // NOTE: Default rules:
  // - parent is pair,elem
  // - state is open

  if (rulespec.parent) {
    if (rulespec.parent.include) {
      matchRule =
        rulespec.parent.include.includes(lex.ctx.rule.parent.name) &&
        (null === matchRule ? true : matchRule)
    }
    if (rulespec.parent.exclude) {
      matchRule =
        !rulespec.parent.exclude.includes(lex.ctx.rule.parent.name) &&
        (null === matchRule ? true : matchRule)
    }
  }

  if (rulespec.current) {
    if (rulespec.current.include) {
      matchRule =
        rulespec.current.include.includes(lex.ctx.rule.name) &&
        (null === matchRule ? true : matchRule)
    }
    if (rulespec.current.exclude) {
      matchRule =
        !rulespec.current.exclude.includes(lex.ctx.rule.name) &&
        (null === matchRule ? true : matchRule)
    }
  }

  // '': don't check, 'oc'|'c'|'o' check, default 'o'
  let rulestate = '' === rulespec.state ? '' : rulespec.state || 'o'
  if (rulestate) {
    matchRule =
      rulestate.includes(lex.ctx.rule.state) &&
      (null === matchRule ? true : matchRule)
  }

  let matchFixed = true
  let fixed = start.fixed
  if (matchRule && null != fixed) {
    matchFixed = false

    fixed = Array.isArray(fixed) ? fixed : [fixed]
    for (let fI = 0; !matchFixed && fI < fixed.length; fI++) {
      if (src.substring(hvpnt.sI).startsWith(fixed[fI])) {
        matchFixed = true

        if (false !== start.consume) {
          if (
            !Array.isArray(start.consume) ||
            start.consume.includes(fixed[fI])
          ) {
            let endI = hvpnt.sI + fixed[fI].length
            for (let fsI = hvpnt.sI; fsI < endI; fsI++) {
              sI++
              cI++
              if ('\n' === src[fsI]) {
                rI++
                cI = 0
              }
            }
          }
        }

        break
      }
    }
  }

  if (matchRule && matchFixed) {
    let startsrc = src.substring(hvpnt.sI, sI)

    if (false !== block.trim) {
      startsrc = startsrc.trim()
    }

    hvpnt.sI = sI
    hvpnt.rI = rI
    hvpnt.cI = cI

    return {
      match: true,
      start: startsrc,
    }
  } else {
    return { match: false }
  }
}

function parseToEnd(
  lex: Lex,
  hvpnt: Point,
  block: any,
  cfg: Config,
): {
  done: boolean
  val: string
  bad?: Token
} {
  let valc = []

  let src = lex.src

  let endspec = block.end
  let fixed: string[] = endspec.fixed
  fixed = 'string' === typeof fixed ? [fixed] : fixed

  let endchars = fixed.map((end) => end[0])
  let endseqs = fixed.map((end) => end.substring(1))

  let escapeChar = block.escapeChar

  let sI = hvpnt.sI // Current point in src
  let rI = hvpnt.rI // Current row
  let cI = hvpnt.cI // Current column

  let done = false
  let c: string = ''
  let endI = sI
  let endCharIndex = 0

  top: do {
    c = src[sI]

    // Check for end
    if (-1 < (endCharIndex = endchars.indexOf(c))) {
      let tail = endseqs[endCharIndex]

      // EOF
      if (undefined === tail || '' === tail) {
        endI = sI + 1
        done = true
        break top
      }

      // Match tail
      else if (
        'string' === typeof tail &&
        tail === src.substring(sI + 1, sI + 1 + tail.length)
      ) {
        endI = sI + 1 + tail.length
        done = true
        break top
      }

      // // regexp
      // else if (tail.exec && (m = tail.exec(src.substring(sI)))) {
      //   pI = 1 + m[0].length
      //   done = true
      //   break endseq
      // }
      // }

      // break top
    }

    if (escapeChar === c) {
      let replacement = block.escape[src[sI + 1]]

      if (null != replacement) {
        c = replacement
        sI++
        cI++
      } else if (block.allowUnknownEscape) {
        c = block.preserveEscapeChar ? src.substring(sI, sI + 2) : src[sI + 1]
        sI++
      } else {
        return {
          done: false,
          val: '',
          bad: lex.bad('invalid_escape', sI, sI + 1),
        }
      }
    }

    valc.push(c)
    sI++
    cI++
    if ('\n' === c) {
      rI++
      cI = 0
    }
  } while (sI <= src.length)

  if (done) {
    if (false !== endspec.consume) {
      let endfixed = src.substring(sI, endI)
      if (
        !Array.isArray(endspec.consume) ||
        endspec.consume.includes(endfixed)
      ) {
        let esI = sI
        for (; esI < endI; esI++) {
          sI++
          cI++
          if ('\n' === src[esI]) {
            rI++
            cI = 0
          }
        }
      }
    }

    hvpnt.sI = sI
    hvpnt.rI = rI
    hvpnt.cI = cI
  }

  let val: any = valc.join('')

  if (block.trim) {
    val = val.trim()
  }

  if (cfg.value.lex && undefined !== cfg.value.def[val]) {
    val = cfg.value.def[val].val
  }

  return {
    done,
    val,
  }
}

Hoover.defaults = {
  block: {
    // // TODO: normalize with defaults
    // "'''": {
    //   open: "'''",
    //   close: "'''",
    //   indent: true,
    //   trim: true,
    //   doubleEscape: false,
    //   lineReplace: null,
    // },
    // // TODO: FOR TEST
    // // '<<': { close: '>>', indent: false }
  },
  lex: {
    order: 4.5e6, // before string, number
  },
} as HooverOptions

export { parseToEnd, Hoover }

export type { HooverOptions }

File: go/hoover.go
/* Copyright (c) 2021-2026 Richard Rodger, MIT License */


package hoover

import (
	jsonic "github.com/jsonicjs/jsonic/go"
)

const Version = "0.1.1"


// Block defines a hoover block configuration.
type Block struct {
	Start              StartSpec
	End                EndSpec
	Token              string            // Token name, default "#HV"
	EscapeChar         string
	Escape             map[string]string
	AllowUnknownEscape *bool // default true
	PreserveEscapeChar bool
	Trim               bool

	name string
	tin  jsonic.Tin
}

// EndSpec defines how a block ends.
type EndSpec struct {
	Fixed   []string // End delimiter(s)
	Consume any      // bool or []string; nil = true
}

// Options configures the Hoover plugin.
type Options struct {
	Block  map[string]*Block
	Action jsonic.AltAction
}

// parseResult is the result of parsing to a block end.
type parseResult struct {
	done bool
	val  any
	err  string
}

// HooverRuleFilter defines include/exclude lists for rule matching.
type HooverRuleFilter struct {
	Include []string
	Exclude []string
}

// HooverRuleSpec defines rule context conditions for matching.
type HooverRuleSpec struct {
	Parent  *HooverRuleFilter
	Current *HooverRuleFilter
	State   string // "" = don't check, "o"/"c"/"oc" = check; default "o"
}

// StartSpec defines how a block starts.
type StartSpec struct {
	Fixed   []string  // Start delimiter(s)
	Consume *bool     // Whether to consume the start delimiter (nil = true)
	Rule    *HooverRuleSpec // Rule context matching
}

// startResult is the result of matching a block start.
type startResult struct {
	match bool
	start string
}

func (b *Block) allowUnknown() bool {
	return b.AllowUnknownEscape == nil || *b.AllowUnknownEscape
}

// Make creates a Hoover plugin with the given options.
// Usage: j.Use(hoover.Make(hoover.Options{...}))
func Make(hopts Options) jsonic.Plugin {
	return func(j *jsonic.Jsonic, _ map[string]any) {
		var blocks []*Block
		tokenMap := map[string]jsonic.Tin{}

		for name, block := range hopts.Block {
			block.name = name
			if block.Token == "" {
				block.Token = "#HV"
			}

			tin := j.Token(block.Token)
			block.tin = tin

			if _, exists := tokenMap[block.Token]; !exists {
				localTin := tin
				j.Rule("val", func(rs *jsonic.RuleSpec) {
					rs.PrependOpen(&jsonic.AltSpec{
						S: [][]jsonic.Tin{{localTin}},
						A: hopts.Action,
					})
				})
			}
			tokenMap[block.Token] = tin

			blocks = append(blocks, block)
		}

		makeHooverMatcher := func(cfg *jsonic.LexConfig, _opts *jsonic.Options) jsonic.LexMatcher {
			return func(lex *jsonic.Lex, rule *jsonic.Rule) *jsonic.Token {
				for _, block := range blocks {
					pnt := lex.Cursor()

					// When no start delimiter is defined (rule-context-only matching),
					// defer to built-in matchers for chars they would handle:
					// strings, numbers, fixed tokens, spaces, lines, comments.
					// This emulates the TS behavior where hoover runs after those
					// matchers in priority order.
					if block.Start.Fixed == nil && pnt.SI < pnt.Len {
						c := rune(lex.Src[pnt.SI])
						if cfg.StringChars[c] || cfg.FixedTokens[string(c)] != 0 ||
							cfg.SpaceChars[c] || cfg.LineChars[c] ||
							isNumberStart(c) || isCommentStart(lex.Src, pnt.SI, cfg) {
							continue
						}
					}

					hvpnt := &jsonic.Point{
						Len: pnt.Len,
						SI:  pnt.SI,
						RI:  pnt.RI,
						CI:  pnt.CI,
					}

					sr := matchStart(lex, hvpnt, block)

					if sr.match {
						result := parseToEnd(lex, hvpnt, block, cfg)

						if result.done {
							src := lex.Src[pnt.SI:hvpnt.SI]
							tkn := lex.Token(block.Token, block.tin, result.val, src)

							pnt.SI = hvpnt.SI
							pnt.RI = hvpnt.RI
							pnt.CI = hvpnt.CI

							return tkn
						}
					}
				}
				return nil
			}
		}

		j.MergeOptions(jsonic.Options{
			Lex: &jsonic.LexOptions{
				Match: map[string]*jsonic.MatchSpec{
					"hoover": {
						Order: 1500000,
						Make:  makeHooverMatcher,
					},
				},
			},
		})
	}
}

func matchStart(
	lex *jsonic.Lex,
	hvpnt *jsonic.Point,
	block *Block,
) startResult {
	src := lex.Src
	rule := lex.Ctx.Rule

	sI := hvpnt.SI
	rI := hvpnt.RI
	cI := hvpnt.CI

	start := block.Start
	rulespec := start.Rule

	// Rule context matching
	var matchRule *bool
	setMatch := func(val bool) {
		if matchRule == nil {
			t := val
			matchRule = &t
		} else {
			t := val && *matchRule
			matchRule = &t
		}
	}

	if rulespec != nil {
		if rulespec.Parent != nil {
			if rule == nil || rule.Parent == nil {
				return startResult{match: false}
			}
			if rulespec.Parent.Include != nil {
				setMatch(containsStr(rulespec.Parent.Include, rule.Parent.Name))
			}
			if rulespec.Parent.Exclude != nil {
				setMatch(!containsStr(rulespec.Parent.Exclude, rule.Parent.Name))
			}
		}

		if rulespec.Current != nil {
			if rule == nil {
				return startResult{match: false}
			}
			if rulespec.Current.Include != nil {
				setMatch(containsStr(rulespec.Current.Include, rule.Name))
			}
			if rulespec.Current.Exclude != nil {
				setMatch(!containsStr(rulespec.Current.Exclude, rule.Name))
			}
		}
	}

	// Rule state check: default "o" (open)
	rulestate := "o"
	if rulespec != nil {
		if rulespec.State == "" {
			rulestate = ""
		} else {
			rulestate = rulespec.State
		}
	}
	if rulestate != "" {
		if rule == nil {
			return startResult{match: false}
		}
		setMatch(containsChar(rulestate, rule.State))
	}

	if matchRule != nil && !*matchRule {
		return startResult{match: false}
	}

	// Fixed delimiter matching
	matchFixed := true
	fixed := start.Fixed

	if fixed != nil {
		matchFixed = false

		for _, f := range fixed {
			if sI+len(f) <= len(src) && src[sI:sI+len(f)] == f {
				matchFixed = true

				consume := start.Consume == nil || *start.Consume
				if consume {
					endI := sI + len(f)
					for fsI := sI; fsI < endI; fsI++ {
						sI++
						cI++
						if src[fsI] == '\n' {
							rI++
							cI = 1
						}
					}
				}

				break
			}
		}
	}

	if matchFixed {
		startsrc := src[hvpnt.SI:sI]

		if block.Trim {
			startsrc = trimString(startsrc)
		}

		hvpnt.SI = sI
		hvpnt.RI = rI
		hvpnt.CI = cI

		return startResult{
			match: true,
			start: startsrc,
		}
	}

	return startResult{match: false}
}

func parseToEnd(
	lex *jsonic.Lex,
	hvpnt *jsonic.Point,
	block *Block,
	cfg *jsonic.LexConfig,
) parseResult {
	var valc []byte

	src := lex.Src

	endspec := block.End
	fixed := endspec.Fixed

	endchars := make([]byte, len(fixed))
	endseqs := make([]string, len(fixed))
	for i, end := range fixed {
		if len(end) > 0 {
			endchars[i] = end[0]
			endseqs[i] = end[1:]
		} else {
			// Empty string = EOF marker
			endchars[i] = 0
			endseqs[i] = ""
		}
	}

	escapeChar := byte(0)
	if block.EscapeChar != "" {
		escapeChar = block.EscapeChar[0]
	}

	sI := hvpnt.SI
	rI := hvpnt.RI
	cI := hvpnt.CI

	done := false
	endI := sI

	for sI <= len(src) {
		// EOF check
		if sI == len(src) {
			for i, ec := range endchars {
				if ec == 0 && endseqs[i] == "" {
					endI = sI
					done = true
					break
				}
			}
			break
		}

		c := src[sI]

		// Check for end delimiters
		endCharIndex := -1
		for i, ec := range endchars {
			if ec == c {
				endCharIndex = i
				break
			}
		}

		if endCharIndex >= 0 {
			tail := endseqs[endCharIndex]

			if tail == "" {
				// Single char end delimiter
				endI = sI + 1
				done = true
				break
			}

			if sI+1+len(tail) <= len(src) && src[sI+1:sI+1+len(tail)] == tail {
				endI = sI + 1 + len(tail)
				done = true
				break
			}
		}

		// Handle escape sequences
		if escapeChar != 0 && c == escapeChar && sI+1 < len(src) {
			nextChar := string(src[sI+1])
			if block.Escape != nil {
				if replacement, ok := block.Escape[nextChar]; ok {
					valc = append(valc, []byte(replacement)...)
					sI += 2
					cI += 2
					continue
				}
			}
			if block.allowUnknown() {
				if block.PreserveEscapeChar {
					valc = append(valc, c)
				}
				valc = append(valc, src[sI+1])
				sI += 2
				cI += 2
				continue
			}
			return parseResult{
				done: false,
				val:  "",
				err:  "invalid_escape",
			}
		}

		valc = append(valc, c)
		sI++
		cI++
		if c == '\n' {
			rI++
			cI = 1
		}
	}

	if done {
		if shouldConsumeEnd(endspec, src, sI, endI) {
			for esI := sI; esI < endI; esI++ {
				sI++
				cI++
				if src[esI] == '\n' {
					rI++
					cI = 1
				}
			}
		}

		hvpnt.SI = sI
		hvpnt.RI = rI
		hvpnt.CI = cI
	}

	val := string(valc)

	if block.Trim {
		val = trimString(val)
	}

	// Resolve defined values (e.g. "true" -> true, "null" -> nil)
	var result any = val
	if cfg != nil && cfg.ValueLex && cfg.ValueDef != nil {
		if defVal, ok := cfg.ValueDef[val]; ok {
			result = defVal
		}
	}

	return parseResult{
		done: done,
		val:  result,
	}
}

func shouldConsumeEnd(endspec EndSpec, src string, sI, endI int) bool {
	if endspec.Consume == nil {
		return true
	}

	switch v := endspec.Consume.(type) {
	case bool:
		return v
	case []string:
		endfixed := src[sI:endI]
		for _, s := range v {
			if s == endfixed {
				return true
			}
		}
		return false
	}

	return true
}

func containsStr(list []string, s string) bool {
	for _, item := range list {
		if item == s {
			return true
		}
	}
	return false
}

func containsChar(s string, sub string) bool {
	for _, c := range sub {
		for _, sc := range s {
			if c == sc {
				return true
			}
		}
	}
	return false
}

func isNumberStart(c rune) bool {
	return (c >= '0' && c <= '9') || c == '-' || c == '+'
}

func isCommentStart(src string, sI int, cfg *jsonic.LexConfig) bool {
	rest := src[sI:]
	for _, cl := range cfg.CommentLine {
		if len(rest) >= len(cl) && rest[:len(cl)] == cl {
			return true
		}
	}
	for _, cb := range cfg.CommentBlock {
		if len(rest) >= len(cb[0]) && rest[:len(cb[0])] == cb[0] {
			return true
		}
	}
	return false
}

func trimString(s string) string {
	start := 0
	end := len(s)
	for start < end && (s[start] == ' ' || s[start] == '\t' || s[start] == '\r' || s[start] == '\n') {
		start++
	}
	for end > start && (s[end-1] == ' ' || s[end-1] == '\t' || s[end-1] == '\r' || s[end-1] == '\n') {
		end--
	}
	return s[start:end]
}
